File: benchmarks/problem12/scala/main.scala
import scala.collection._

object facben {

  class PrimeGenerator {
    var n: Long = 3L
    var last: Long = 2L
    val sieve: mutable.LongMap[Long] = mutable.LongMap.empty[Long]

    def next(): Long = {
      var prime = 0L
      while ({ prime = sieve.getOrElse(n, 0); prime != 0L }) {
        sieve.remove(n)

        var composite: Long = n + prime + prime
        while (sieve.contains(composite)) {
          composite += prime + prime
        }

        sieve += (composite, prime)
        n += 2L
      }

      sieve += (n * n, n)
      val r = last
      last = n
      n += 2L
      r
    }
  }

  class FactorFinder {
    val known: mutable.LongMap[mutable.Set[Long]] = initKnown()
    val nextPrimes: PrimeGenerator = new PrimeGenerator()
    val knownPrimes: mutable.ArrayBuffer[Long] = mutable.ArrayBuffer.empty[Long]

    def initKnown(): mutable.LongMap[mutable.Set[Long]] = {
      var lm = mutable.LongMap.empty[mutable.Set[Long]]
      var s = mutable.Set.empty[Long]
      s += 1L
      lm += (1L, s)
      lm
    }

    def getPrimeFactors(of: Long): mutable.ArrayBuffer[Long] = {
      var factors = mutable.ArrayBuffer.empty[Long]
      var quotient = of

      val kp = knownPrimes.iterator
      var prime = 0L
      while (kp.hasNext && { prime = kp.next(); true } && prime <= quotient) {

        var remainder = quotient % prime
        while (remainder == 0L) {
          quotient /= prime
          remainder = quotient % prime
          factors += prime
        }
      }

      prime = nextPrimes.next()
      knownPrimes += prime
      while (prime <= quotient) {

        var remainder = quotient % prime
        while (remainder == 0L) {
          quotient /= prime
          remainder = quotient % prime
          factors += prime
        }

        prime = nextPrimes.next()
        knownPrimes += prime
      }

      factors
    }

    def getFactors(of: Long): mutable.Set[Long] = {
      known.get(of) match {
        case Some(facs) => facs
        case _ => {
          val primeFactors = getPrimeFactors(of)
          var factors = mutable.HashSet.empty[Long]
          factors += 1L
          factors += of
          for (prime <- primeFactors) {
            val factor = of / prime
            for (subfactor <- getFactors(factor)) {
              factors += subfactor
            }
          }

          known += (of, factors)
          factors
        }
      }
    }
  }

  def solve(): Long = {

    val finder = new FactorFinder()
    var adder = 0L
    var tn = 0L
    var length = 0

    while (length < 1000) {
      adder += 1L
      tn += adder

      val factors = finder.getFactors(tn)
      length = factors.size
    }

    tn
  }

  def main(args: Array[String]): Unit = {
    val answer = solve()
    println(s"Answer: $answer")
  }
}

File: benchmarks/problem12/julia/main.jl
# run with julia main.jl
# julia version used: 0.6.2



mutable struct PrimeGenerator
    n::Int64
    lastn::Int64
    sieve::Dict{Int64,Int64}
    PrimeGenerator() = new(3, 2, Dict())
end

function nextPrime(self::PrimeGenerator)
    prime = get(self.sieve, self.n, 0)
    while prime != 0

        delete!(self.sieve, self.n)

        composite = self.n + prime + prime
        while haskey(self.sieve, composite)
            composite += prime + prime
        end
        self.sieve[composite] = prime

        self.n += 2
        prime = get(self.sieve, self.n, 0)
    end

    self.sieve[self.n * self.n] = self.n
    r = self.lastn
    self.lastn = self.n
    self.n += 2
    r
end

struct FactorFinder
    known::Dict{Int64,Set{Int64}}
    nextPrimes::PrimeGenerator
    knownPrimes::Array{Int64}
    FactorFinder() = new(Dict(1 => Set(1)), PrimeGenerator(), [])
end

function getPrimeFactors(self::FactorFinder, of::Int64)
    factors = []
    quotient = of

    for prime in self.knownPrimes
        if prime > quotient
            return factors
        end

        remainder = rem(quotient, prime)
        while remainder == 0
            quotient = div(quotient, prime)
            remainder = rem(quotient, prime)
            push!(factors, prime)
        end
    end

    while true
        prime = nextPrime(self.nextPrimes)
        push!(self.knownPrimes, prime)

        if prime > quotient
            return factors
        end

        remainder = rem(quotient, prime)
        while remainder == 0
            quotient = div(quotient, prime)
            remainder = rem(quotient, prime)
            push!(factors, prime)
        end
    end

    factors
end

function getFactors(self::FactorFinder, of::Int64)
    existing = get(self.known, of, 0)
    if existing != 0
        return existing
    end

    factors = Set{Int64}([1, of])
    for prime in getPrimeFactors(self, of)
        factor = div(of, prime)
        for subfactor in getFactors(self, factor)
            push!(factors, subfactor)
        end
    end

    self.known[of] = factors
    factors
end

function solve()
    finder = FactorFinder()
    adder = 0
    tn = 0

    while true
        adder += 1
        tn += adder

        factors = getFactors(finder, tn)
        if length(factors) > 1000
            return tn
        end
    end
end

answer = solve()
println("Answer: " * string(answer))